### PR TITLE
Add ability to ignore pattern when formatting

### LINKED ISSUE #421: Allow file path(s) to be excluded from cadl format
I'm currently doing `cadl format -c **/*.cadl` however this is picking up on an additional cadl file included via the node modules that I don't actually have control over (resource.cadl). It's simple enough to run cadl format locally but this doesn't work if I want to enforce formatting in the build pipeline.

```
PS C:\redacted-path> cadl format -c **/*.cadl
Cadl compiler v0.29.0

Found 3 unformatted files:
 - src/path/cadl/main.cadl
 - src/path/cadl/node_modules/@cadl-lang/rest/lib/resource.cadl
 - src/path/cadl/foo.cadl
```

### PR DESCRIPTION
**Pull request by timotheeguerin**
fix #421 

Adds the ability to exclude pattern from being formatted. It auto exclude `node_modules`. 

Usage: 
```
# Exclude certain patterns. Either use `!` prefix or pass it via the `--exclude` or `-x` option.
cadl format **/*.cadl "!mytestfolder/**/*"
cadl format **/*.cadl --exclude "mytestfolder/**/*"
```

Updated to use `globby` instead of `glob` which is a wrapper on top of `fast-glob` but provide ignore functionality and a promise interface.

Had to update the node types as there was an error in the globby definition with `import from "node:url"` not found.

### CODE REVIEW

File: packages/cadl-vscode/package.json [hunk 1]
```diff
@@ -109,7 +109,7 @@
     "@rollup/plugin-node-resolve": "~13.1.3",
     "@types/mkdirp": "~1.0.1",
     "@types/mocha": "~9.1.0",
-    "@types/node": "~14.0.27",
+    "@types/node": "~17.0.23",
```
Comment (nguerrera) [inline packages/cadl-vscode/package.json:112]: Shouldn't we keep this at the minimum version we support? Does 16.0.x work? Could we wait for the Node 14 support removal to go in to adress the issue here?

Comment (timotheeguerin) [reply packages/cadl-vscode/package.json:112]: Yeah I guess I just took the latest version, should probably work with node 16

File: packages/compiler/package.json [hunk 2]
```diff
@@ -68,12 +68,11 @@
     "change-case": "~4.1.2"
   },
   "devDependencies": {
-    "@types/glob": "~7.1.3",
     "@types/js-yaml": "~4.0.1",
     "@types/mkdirp": "~1.0.1",
     "@types/mocha": "~9.1.0",
     "@types/mustache": "~4.1.2",
-    "@types/node": "~14.0.27",
+    "@types/node": "~16.11.7",
```
Comment (nguerrera) [inline packages/compiler/package.json:75]: I think it's best to keep 16.0.x so we don't take a dependency on an API added in minor release and we can keep our requirment just "node 16". Basically, you can revert this after merging with main.

Comment (timotheeguerin) [reply packages/compiler/package.json:75]: sounds good, just taking up your change instead of this.

**Review by nguerrera — APPROVED** [review]

**Pull request merged**